### DISCUSSION TITLE
Classification Example: Camera only capturing once when using custom CNN

### DISCUSSION THREAD

**Discussion by Jool99** [bitcraze/discussions#1220] [Q&A]
In the classification example, I exchanged the tflite file for my own. Then the cam_handler is only called once. The inference is successful the first time. What could this depend on?

All answers are appreciated. Thanks in advance! :)

**Comment by Jool99** [comment]
The problem seems to be timing related. The new network is really small. Adding pi_time_wait_us(1000000) in the camera callback made it work.

This small two layer network is just for testing something and maybe this will never be a problem with a larger network worth running. But it seems unstable to be depending on delays.

I am still curious to why this is happening and how to make it robust.

**Reply by gemenerik** [reply]
Hi Jool99, can you share your model architecture or even better your `.tflite` file?

**Comment by Jool99** [comment]
Here is the model: [simplemodel.zip](https://github.com/bitcraze/discussions/files/14522758/simplemodel.zip)

I think it can also be reproduced by commenting out this line:

`pi_cluster_send_task_to_cl(&cluster_dev, task);`

**Reply by gemenerik** [reply]
This line is actually triggering inference of the model, so by commenting it out the CNN never runs.

**Reply by gemenerik** [reply]
I'm not sure, but the problem might be that your model has a different shape output: int8[1,1] vs uint8[1,2] for the supplied model. 
I like [netron](netron.app) for inspecting models.

Also note the signed vs unsigned. This shouldn't prevent the net from running but can give you unexpected results.

**Reply by Jool99** [reply]
I was wondering about the signed vs unsigned. Will try netron. Thank you!

**Comment by Jool99** [comment]
> This line is actually triggering inference of the model, so by commenting it out the CNN never runs.

Yes but the problem was that the camera is not capturing a second picture, or at least not calling the registered callback function cam_handler, if there is not sufficient delay before calling pi_camera_capture_async the second time. The delay can be introduced by running a sufficiently large CNN or replacing the CNN execution with a delay. I have only tried 1 second.

Maybe this is the reason for the comment bellow in main.c in the face detection example?
 `//Set Cluster Frequency to 75MHz - max (175000000) leads to more camera failures (unknown why)`

Using a max waiting time and restarting the camera if the callback function registered with pi_camera_capture_async has not been called yet like in the face detection demo is probably the answer to this thread.

**Reply by gemenerik** [reply]
> ... the problem was that the camera is not capturing a second picture, or at least not calling the registered callback function cam_handler

Ah right! Yes, I can reproduce this. At first I thought it might have to do with exceeding the maximum camera rates (60 FPS@QVGA, 120 FPS@QQVGA) but after experimenting noticed that I already get down to ~16FPS with a 0.002 second delay in the camera capture loop. Although the required delay to get it to run is very small (somewhere between 0.001 and 0.002 seconds) and has little effect on the already not-so-speedy capture loop it is peculiar that it requires any further delay at all... I think it's best to open an issue [here](https://github.com/GreenWaves-Technologies/gap_sdk)

**Reply by Jool99** [reply]
Thanks for ur time! :) Made an [issue](https://github.com/GreenWaves-Technologies/gap_sdk/issues/411)
